File: ios/RiveLog.swift
enum RiveLogLevel: Int, Comparable {
    case debug = 0
    case info = 1
    case warn = 2
    case error = 3

    static func < (lhs: RiveLogLevel, rhs: RiveLogLevel) -> Bool {
        lhs.rawValue < rhs.rawValue
    }

    init?(string: String) {
        switch string {
        case "debug": self = .debug
        case "info": self = .info
        case "warn": self = .warn
        case "error": self = .error
        default: return nil
        }
    }
}

enum RiveLog {
    static var handler: ((String, String, String) -> Void)?
    static var minLevel: RiveLogLevel = .warn

    static func e(_ tag: String, _ message: String) {
        guard .error >= minLevel else { return }
        if let handler = handler {
            handler("error", tag, message)
        } else {
            RCTLogError("[\(tag)] \(message)")
        }
    }

    static func w(_ tag: String, _ message: String) {
        guard .warn >= minLevel else { return }
        if let handler = handler {
            handler("warn", tag, message)
        } else {
            RCTLogWarn("[\(tag)] \(message)")
        }
    }

    static func i(_ tag: String, _ message: String) {
        guard .info >= minLevel else { return }
        if let handler = handler {
            handler("info", tag, message)
        } else {
            RCTLogInfo("[\(tag)] \(message)")
        }
    }

    static func d(_ tag: String, _ message: String) {
        guard .debug >= minLevel else { return }
        if let handler = handler {
            handler("debug", tag, message)
        } else {
            RCTLog("[\(tag)] \(message)")
        }
    }
}

File: android/src/main/java/com/margelo/nitro/rive/RiveLog.kt
package com.margelo.nitro.rive

import android.util.Log

enum class RiveLogLevel(val priority: Int) {
    DEBUG(0),
    INFO(1),
    WARN(2),
    ERROR(3),
    ;

  companion object {
        fun fromString(level: String): RiveLogLevel? = when (level) {
            "debug" -> DEBUG
            "info" -> INFO
            "warn" -> WARN
            "error" -> ERROR
            else -> null
        }
    }
}

object RiveLog {
    var handler: ((String, String, String) -> Unit)? = null
    var minLevel: RiveLogLevel = RiveLogLevel.WARN

    fun e(tag: String, message: String) {
        if (RiveLogLevel.ERROR.priority < minLevel.priority) return
        handler?.invoke("error", tag, message) ?: Log.e(tag, message)
    }

    fun w(tag: String, message: String) {
        if (RiveLogLevel.WARN.priority < minLevel.priority) return
        handler?.invoke("warn", tag, message) ?: Log.w(tag, message)
    }

    fun i(tag: String, message: String) {
        if (RiveLogLevel.INFO.priority < minLevel.priority) return
        handler?.invoke("info", tag, message) ?: Log.i(tag, message)
    }

    fun d(tag: String, message: String) {
        if (RiveLogLevel.DEBUG.priority < minLevel.priority) return
        handler?.invoke("debug", tag, message) ?: Log.d(tag, message)
    }
}
